daemon/errors.go

```diff
@@ -1,53 +1,112 @@
 package daemon
 
 import (
-	"fmt"
-
-	"github.com/docker/docker/api/errors"
+	"github.com/pkg/errors"
 )
 
-func (d *Daemon) imageNotExistToErrcode(err error) error {
-	if dne, isDNE := err.(ErrImageDoesNotExist); isDNE {
-		return errors.NewRequestNotFoundError(dne)
-	}
-	return err
+func errNotRunning(id string) error {
+	return stateConflictError{errors.Errorf("Container %s is not running", id)}
 }
 
-type errNotRunning struct {
-	containerID string
+type notFoundError struct {
+	err error
 }
 
-func (e errNotRunning) Error() string {
-	return fmt.Sprintf("Container %s is not running", e.containerID)
+func (e notFoundError) Error() string {
+	return e.err.Error()
 }
 
-func (e errNotRunning) ContainerIsRunning() bool {
-	return false
+func (e notFoundError) NotFound() {}
+
+func (e notFoundError) Cause() error {
+	return e.err
 }
 
+type stateConflictError struct {
+	err error
+}
+
+func (e stateConflictError) Error() string {
+	return e.err.Error()
+}
+
+func (e stateConflictError) Cause() error {
+	return e.err
+}
+
+func (e stateConflictError) Conflict() {}
+
 func errContainerIsRestarting(containerID string) error {
-	err := fmt.Errorf("Container %s is restarting, wait until the container is running", containerID)
-	return errors.NewRequestConflictError(err)
+	err := errors.Errorf("Container %s is restarting, wait until the container is running", containerID)
+	return stateConflictError{err}
 }
 
 func errExecNotFound(id string) error {
-	err := fmt.Errorf("No such exec instance '%s' found in daemon", id)
-	return errors.NewRequestNotFoundError(err)
+	err := errors.Errorf("No such exec instance '%s' found in daemon", id)
+	return notFoundError{err}
 }
 
 func errExecPaused(id string) error {
-	err := fmt.Errorf("Container %s is paused, unpause the container before exec", id)
-	return errors.NewRequestConflictError(err)
+	err := errors.Errorf("Container %s is paused, unpause the container before exec", id)
+	return notFoundError{err}
 }
 
-type errNotFound struct {
-	containerID string
+func containerNotFound(id string) error {
+	return notFoundError{errors.Errorf("Container %s is not found", id)}
 }
 
-func (e errNotFound) Error() string {
-	return fmt.Sprintf("Container %s is not found", e.containerID)
+type validationError struct {
+	err error
 }
 
-func (e errNotFound) ContainerNotFound() bool {
-	return true
+func (e validationError) Error() string {
+	return e.err.Error()
+}
+
+func (e validationError) InvalidParameter() {}
+
+func (e validationError) Cause() error {
+	return e.err
+}
+
+type notAllowedError struct {
+	err error
+}
+
+func (e notAllowedError) Error() string {
+	return e.err.Error()
+}
+
+func (e notAllowedError) Forbidden() {}
+
+func (e notAllowedError) Cause() error {
+	return e.err
+}
+
+type notModifiedError struct {
+	err error
+}
+
+func (e notModifiedError) Error() string {
+	return e.err.Error()
```

Comment:
This struct could have a flag indicating whether the container was already started or already stopped, and generate the error message in the `Error` method.